File: v1/scala/acronym/src/main/scala/Acronym.scala
case class Acronym(phrase: String) {
  def acronymChar(str: String) = str(0).toUpper

  val pattern = "[A-Z]+[a-z]*|[a-z]+".r
  val words = pattern.findAllIn(phrase)
  val abbreviate = words.map(acronymChar).mkString
}

File: haskell/acronym/src/Acronym.hs
module Acronym (abbreviate) where

import Data.Char (toUpper, isUpper)
import Data.List (map, concatMap, filter)
import Data.List.Split (splitOn)

wordToLetters :: String -> String
wordToLetters [] = []
wordToLetters (x:xs) = if all isUpper xs then [toUpper x] else toUpper x:filter isUpper xs

phraseToWords :: String -> [String]
phraseToWords phrase = concatMap (splitOn "-") $ words phrase

abbreviate :: String -> String
abbreviate phrase = concatMap wordToLetters $ phraseToWords phrase
